doc/xpdf.1
.\" Copyright 1996-2014 Glyph & Cog, LLC
.\" Copyright 2014-2020 Adam Sampson <ats@offog.org>
.TH xpdf 1 "24 Dec 2020"
.SH NAME
xpdf \- Portable Document Format (PDF) file viewer for X (xpopple)
.SH SYNOPSIS
.B xpdf
[options]
.RI [ PDF-file
.RI [ page " | +" dest ]]
.SH DESCRIPTION
.B xpopple
is a viewer for Portable Document Format (PDF) files.  (These are also
sometimes also called 'Acrobat' files, from the name of Adobe's PDF
software.)  xpopple runs under the X Window System on UNIX.
.PP
xpopple uses the Poppler library to process PDF files.  Poppler was originally
derived from version 3 of Glyph & Cog, LLC's Xpdf suite, but Poppler doesn't
include the Xpdf viewer program itself.  xpopple is the Xpdf program, modified
to use Poppler.
.PP
To run xpopple, simply type:
.PP
.RS
xpdf file.pdf
.RE
.PP
where
.I file.pdf
is your PDF file.  The file name can be followed by a number
specifying the page which should be displayed first, e.g.:
.PP
.RS
xpdf file.pdf 18
.RE
.PP
You can also give a named destination, prefixed with '+' in place of
the page number.  (This is only useful with PDF files that provide
named destination targets.)
.PP
You can also start xpdf without opening any files:
.PP
.RS
xpdf
.RE
.PP
.SH CONFIGURATION FILE
Xpdf reads a configuration file at startup.  It first tries to find
the user's private config file, ~/.xpdfrc.  If that doesn't exist, it
looks for a system-wide config file, /etc/xpdf/xpdfrc.  See the
.BR xpdfrc (5)
man page for details.
.SH OPTIONS
Many of the following options can be set with configuration file
commands or X resources.  These are listed in square brackets with the
description of the corresponding command line option.
.TP
.BI \-g " geometry"
Set the initial window geometry.
.RB ( \-geometry
is equivalent.)
.RB "[X resource: " Xpdf.geometry ]
.TP
.BI \-title " title"
Set the window title.  By default, the title will be "xpdf: foo.pdf".
.RB "[X resource: " Xpdf.title ]
.TP
.B \-cmap
Install a private colormap.  This is ignored on TrueColor visuals.
.RB "[X resource: " Xpdf.installCmap ]
.TP
.BI \-rgb " number"
Set the size of largest RGB cube xpdf will try to allocate.  The
default is 5 (for a 5x5x5 cube); set to a smaller number to conserve
color table entries.  This is ignored with private colormaps and on
TrueColor visuals.
.RB "[X resource: " Xpdf.rgbCubeSize ]
.TP
.B \-rv
Set reverse video mode.  This reverses the colors of everything except
images.  It may not always produce great results for PDF files which
do weird things with color.  This also causes the paper color to
default to black.
.RB "[X resource: " Xpdf.reverseVideo ]
.TP
.BI \-papercolor " color"
Set the "paper color", i.e., the background of the page display.  This
will not work too well with PDF files that do things like filling in
white behind the text.
.RB "[X resource: " Xpdf.paperColor ]
.TP
.BI \-mattecolor " color"
Set the matte color, i.e., the color used for background outside the
actual page area.  (There is a separate setting,
Xpdf.fullScreenMatteColor, for full-screen mode.)
.RB "[X resource: " Xpdf.matteColor ]
.TP
.BI \-z " zoom"
Set the initial zoom factor.  A number specifies a zoom percentage,
where 100 means 72 dpi.  You may also specify \'page', to fit the page
to the window size, or \'width', to fit the page width to the window
width, or \'height', to fit the page height to the window height
.RB "[config file: " initialZoom "; or X resource: " Xpdf.initialZoom ]
.TP
.B \-cont
Start in continuous view mode, i.e., with one vertical scroll bar for
the whole document.
.RB "[config file: " continuousView ]
.TP
.BI \-ps " PS-file"
Set the default file name for PostScript output (i.e., the name which
will appear in the print dialog).  This can also be of the
form '| command' to pipe the PostScript through a command.
.RB "[config file: " psFile ]
.TP
.BI \-paper " size"
Set the paper size to one of "letter", "legal", "A4", or "A3".  This
can also be set to "match", which will set the paper size to match the
size specified in the PDF file.
.RB "[config file: " psPaperSize ]
.TP
.BI \-paperw " size"
Set the paper width, in points.
.RB "[config file: " psPaperSize ]
.TP
.BI \-paperh " size"
Set the paper height, in points.
.RB "[config file: " psPaperSize ]
.TP
.B \-level1
Generate Level 1 PostScript.  The resulting PostScript files will be
significantly larger (if they contain images), but will print on Level
1 printers.  This also converts all images to black and white.
.RB "[config file: " psLevel ]
.TP
.BI \-enc " encoding-name"
Sets the encoding to use for text output.  The
.I encoding\-name
must be defined with the unicodeMap command (see
.BR xpdfrc (5)).
This defaults to "Latin1" (which is a built-in encoding).
.RB "[config file: " textEncoding ]
.TP
.BI \-opw " password"
Specify the owner password for the PDF file.  Providing this will
bypass all security restrictions.
.TP
.BI \-upw " password"
Specify the user password for the PDF file.
.TP
.B \-fullscreen
Open xpdf in full-screen mode, useful for presentations.
.TP
.BI \-pagecmd " command"
Run a shell command each time a new page is displayed, with the page
number as an argument.  This can be used for presentation recording.
.RB "[config file: " pageCommand ]
.TP
.BI \-remote " name"
Start/contact xpdf remote server with specified name (see the
.B "REMOTE SERVER MODE"
section below).
.TP
.BI \-exec " command"
Execute a command (see the
.B COMMANDS
section below) in an xpdf remote server window (with \-remote only).
.TP
.B \-reload
Reload xpdf remote server window (with \-remote only).
.TP
.B \-raise
Raise xpdf remote server window (with \-remote only).
.TP
.B \-quit
Kill xpdf remote server (with \-remote only).
.TP
.B \-cmd
Print commands as they're executed (useful for debugging).
.RB "[config file: " printCommands ]
.TP
.B \-q
Don't print any messages or errors.
.RB "[config file: " errQuiet ]
.TP
.BI \-cfg " config-file"
Read
.I config-file
in place of ~/.xpdfrc or the system-wide config file.
.TP
.B \-v
Print copyright and version information.
.TP
.B \-h
Print usage information.
.RB ( \-help ,
.B \-\-help
and
.B \-?
are equivalent.)
.PP
Several other standard X options and resources will work as expected:
.TP
.BI \-display " display"
.RB "[X resource: " Xpdf.display ]
.TP
.BI \-fg " color"
.RB ( \-foreground
is equivalent.)
.RB "[X resource: " xpdf*Foreground ]
.TP
.BI \-bg " color"
.RB ( \-background
is equivalent.)
.RB "[X resource: " xpdf*Background ]
.TP
.BI \-font " font"
.RB ( \-fn
is equivalent.)
.RB "[X resource: " xpdf*fontList ]
.PP
The color and font options only affect the user interface elements,
not the PDF display (the 'paper').
.PP
The following X resources do not have command line option equivalents:
.TP
.B Xpdf.toolTipEnable
Enables (if set to true) or disables (if set to false) the tool-tips
on the toolbar buttons.
.TP
.B Xpdf.fullScreenMatteColor
Sets the matte color to be used in full-screen mode.  The default
setting is "black".
.SH OBSOLETE OPTIONS
The following command line options were accepted by earlier versions of Xpdf
but are no longer allowed.  In most cases, this is because they control
rendering options that Poppler does not support.
.TP
.BI \-aa " yes | no"
Obsolete.  Controlled the use of font anti-aliasing in the PDF rasterizer.
.TP
.BI \-aaVector " yes | no"
Obsolete.  Controlled the use of vector anti-aliasing in the PDF rasterizer.
.TP
.BI \-eol " unix | dos | mac"
Obsolete.  Set the end-of-line convention to use for text output.
.TP
.BI \-freetype " yes | no"
Obsolete.  Controlled whether FreeType was used for font rasterization.
.TP
.BI \-t1lib " yes | no"
Obsolete.  Controlled whether t1lib was used for font rasterization.
.SH CONTROLS
.SS On-screen controls, at the bottom of the xpdf window
.TP
.B "left/right arrow buttons"
Move to the previous/next page.
.TP
.B "double left/right arrow buttons"
Move backward or forward by ten pages.
.TP
.B "dashed left/right arrow buttons"
Move backward or forward along the history path.
.TP
.B "'Page' entry box"
Move to a specific page number.  Click in the box to activate it, type
the page number, then hit return.
.TP
.B "zoom popup menu"
Change the zoom factor (see the description of the \-z option above).
.TP
.B "binoculars button"
Find a text string.
.TP
.B "print button"
Bring up a dialog for generating a PostScript file.  The dialog has
options to set the pages to be printed and the PostScript file name.
The file name can be '\-' for stdout or '| command' to pipe the
PostScript through a command, e.g., '| lpr'.
.TP
.B "'?' button"
Bring up the 'about xpdf' window.
.TP
.B "link info"
The space between the '?' and 'Quit' buttons is used to show the URL
or external file name when the mouse is over a link.
.TP
.B "'Quit' button"
Quit xpdf.
.PP
.SS Menu
Pressing the right mouse button will post a popup menu with the
following commands:
.TP
.B "Open..."
Open a new PDF file via a file requester.
.TP
.B "Open in new window..."
Create a new window and open a new PDF file via a file requester.
.TP
.B "Reload"
Reload the current PDF file.  Note that Xpdf will reload the file
automatically (on a page change or redraw) if it has changed since it
was last loaded.
.TP
.B "Save as..."
Save the current file via a file requester.
.TP
.B "Continuous view"
Toggles between single page and continuous view modes.
.TP
.B "Rotate counterclockwise"
Rotate the page 90 degrees counterclockwise.
.TP
.B "Rotate clockwise"
Rotate the page 90 degrees clockwise.  The two rotate commands are
intended primarily for PDF files where the rotation isn't correctly
specified in the file.
.TP
.B "Zoom to selection"
Zoom in to the currently selected rectangle.
.TP
.B "Close"
Close the current window.  If this is the only open window, the
document is closed, but the window is left open (i.e., this menu
command won't quit xpdf).
.TP
.B "Quit"
Quit xpdf.
.PP
.SS Outline
If the PDF contains an outline (a.k.a., bookmarks), there will be an
outline pane on the left side of the window.  The width of the outline
pane is adjustable with a vertical split bar via the knob near its
bottom end.
.PP
.SS Text selection
Dragging the mouse with the left button held down will highlight an
arbitrary rectangle.  Any text inside this rectangle will be copied to
the X selection buffer.
.PP
.SS Links
Clicking on a hyperlink will jump to the link's destination.  A link
to another PDF document will make xpdf load that document.  A 'launch'
link to an executable program will display a dialog, and if
you click 'ok', execute the program.  URL links call an external
command (see the
.B WEB BROWSERS
section below).
.PP
.SS Panning
Dragging the mouse with the middle button held down pans the window.
.PP
.SS Key bindings
.TP
.B o
Open a new PDF file via a file requester.
.TP
.B r
Reload the current PDF file.  Note that Xpdf will reload the file
automatically (on a page change or redraw) if it has changed since it
was last loaded.
.TP
.B s
Save the current PDF file via a file requester.
.TP
.B control-L
Redraw the current page.
.TP
.B control-W
Close the current window.
.TP
.BR f " or " control-F " or " /
Find a text string.
.TP
.B control-G
Find next occurrence.
.TP
.B control-V
Find previous occurrence.
.TP
.B control-P
Print.
.TP
.B n
Move to the next page.  Scrolls to the top of the page, unless scroll
lock is turned on.
.TP
.B p
Move to the previous page.  Scrolls to the top of the page, unless
scroll lock is turned on.
.TP
.BR <Space> " or " <PageDown>
Scroll down on the current page; if already at bottom, move to next
page.
.TP
.BR <Backspace> " or " <Delete> " or " <PageUp>
Scroll up on the current page; if already at top, move to previous
page.
.TP
.B v
Move forward along the history path.
.TP
.B b
Move backward along the history path.
.TP
.B <Home>
Scroll to top of current page.
.TP
.B <End>
Scroll to bottom of current page.
.TP
.B control-<Home>
Scroll to first page of document.
.TP
.B control-<End>
Scroll to last page of document.
.TP
.B arrows
Scroll the current page.
.TP
.B [
Rotate the page 90 degrees counterclockwise.
.TP
.B ]
Rotate the page 90 degrees clockwise.
.TP
.B g
Activate the page number text field ("goto page").
.TP
.B 0
Set the zoom factor to 125% (ready for increment and decrement).
.TP
.B +
Zoom in (increment the zoom factor by 1).
.TP
.B -
Zoom out (decrement the zoom factor by 1).
.TP
.B Z
Activate the zoom number text field.
.TP
.B z
Set the zoom factor to 'page' (fit page to window).
.TP
.B w
Set the zoom factor to 'width' (fit page width to window).
.TP
.B h
Set the zoom factor to 'height' (fit page height to window).
.TP
.B alt-F
Toggle full-screen mode.
.TP
.BR ?
Open the 'about' dialog.
.TP
.BR q " or " <Escape>
Quit xpdf.
.SH "WEB BROWSERS"
If you want to run xpdf automatically from netscape or mosaic (and
probably other browsers) when you click on a link to a PDF file, you
need to edit (or create) the files
.I .mime.types
and
.I .mailcap
in your home directory.  In
.I .mime.types
add the line:
.PP
.RS
application/pdf pdf
.RE
.PP
In
.I .mailcap
add the lines:
.PP
.RS
# Use xpdf to view PDF files.
.RE
.RS
application/pdf; xpdf \-q %s
.RE
.PP
Make sure that xpdf is on your executable search path.
.PP
When you click on a URL link in a PDF file, xpdf will execute the
command specified by the urlCommand config file option, replacing an
occurrence of '%s' with the URL.  For example, to call netscape with
the URL, add this line to your config file:
.PP
.RS
urlCommand "netscape \-remote 'openURL(%s)'"
.RE
.SH COMMANDS
Xpdf's key and mouse bindings are user-configurable, using the bind and
unbind options in the config file (see
.BR xpdfrc (5)).
The bind command allows you to bind a key or mouse button to a
sequence of one or more commands.
.SS Available Commands
The following commands are supported:
.TP
.BI gotoPage( page )
Go to the specified page.
.TP
.BI gotoPageNoScroll( page )
Go to the specified page, with the current relative scroll position.
.TP
.BI gotoDest( dest )
Go to a named destination.
.TP
.B gotoLastPage
Go to the last page in the PDF file.
.TP
.B gotoLastPageNoScroll
Go to the last page in the PDF file, with the current relative scroll
position.
.TP
.B nextPage
Go to the next page.
.TP
.B nextPageNoScroll
Go to the next page, with the current relative scroll position.
.TP
.B prevPage
Go to the previous page.
.TP
.B prevPageNoScroll
Go to the previous page, with the current relative scroll position.
.TP
.B pageUp
Scroll up by one screenful.
.TP
.B pageDown
Scroll down by one screenful.
.TP
.BI scrollLeft( n )
Scroll left by
.I n
pixels.
.TP
.BI scrollRight( n )
Scroll right by
.I n
pixels.
.TP
.BI scrollUp( n )
Scroll up by
.I n
pixels.
.TP
.BI scrollDown( n )
Scroll down by
.I n
pixels.
.TP
.BI scrollUpPrevPage( n )
Scroll up by
.I n
pixels, moving to the previous page if appropriate.
.TP
.BI scrollDownNextPage( n )
Scroll down by
.I n
pixels, moving to the next page if appropriate.
.TP
.B scrollToTopEdge
Scroll to the top edge of the current page, with no horizontal
movement.
.TP
.B scrollToBottomEdge
Scroll to the bottom edge of the current page, with no horizontal
movement.
.TP
.B scrollToLeftEdge
Scroll to the left edge of the current page, with no vertical
movement.
.TP
.B scrollToRightEdge
Scroll to the right edge of the current page, with no vertical
movement.
.TP
.B scrollToTopLeft
Scroll to the top-left corner of the current page.
.TP
.B scrollToBottomRight
Scroll to the bottom-right corner of the current page.
.TP
.B goForward
Move forward along the history path.
.TP
.B goBackward
Move backward along the history path.
.TP
.BI zoomPercent( z )
Set the zoom factor to
.IR z %.
.TP
.B zoomFitPage
Set the zoom factor to fit-page.
.TP
.B zoomFitWidth
Set the zoom factor to fit-width.
.TP
.B zoomFitHeight
Set the zoom factor to fit-height.
.TP
.B zoomIn
Zoom in - go to the next higher zoom factor.
.TP
.B zoomOut
Zoom out - go the next lower zoom factor.
.TP
.B zoomToSelection
Zoom in to the currently selected rectangle, if there is one.
.TP
.B rotateCW
Rotate the page 90 degrees clockwise.
.TP
.B rotateCCW
Rotate the page 90 degrees counterclockwise.
.TP
.BI setSelection( pg , ulx , uly , lrx , lry )
Set the selection to the specified coordinates on the specified page.
.TP
.B continuousMode
Go to continuous view mode.
.TP
.B singlePageMode
Go to single-page view mode.
.TP
.B toggleContinuousMode
Toggle between continuous and single page view modes.
.TP
.B fullScreenMode
Go to full-screen mode.
.TP
.B windowMode
Go to window (non-full-screen) mode.
.TP
.B toggleFullScreenMode
Toggle between full-screen and window modes.
.TP
.B open
Open a PDF file in this window, using the open dialog.
.TP
.B openInNewWin
Open a PDF file in a new window, using the open dialog.
.TP
.BI openFile( file )
Open a specified PDF file in this window.
.TP
.BI openFileInNewWin( file )
Open a specified PDF file in a new window.
.TP
.BI openFileAtDest( file , dest )
Open a specified PDF file in this window and go to a named
destination.
.TP
.BI openFileAtDestInNewWin( file , dest )
Open a specified PDF file in a new window and go to a named
destination.
.TP
.BI openFileAtPage( file , page )
Open a specified PDF file in this window and go to a page.
.TP
.BI openFileAtPageInNewWin( file , page )
Open a specified PDF file in a new window and go to a page.
.TP
.B reload
Reload the current PDF file.
.TP
.B saveAs
Save the PDF file, using the 'save as' dialog.
.TP
.B redraw
Redraw the window.
.TP
.B raise
Raise the window to the front.
.TP
.B closeWindow
Close the window.  If this was the last open window, clear the window,
but don't quit from Xpdf.
.TP
.B closeWindowOrQuit
Close the window.  If this was the last open window, quit from Xpdf.
.TP
.BI run( external-command-string )
Run an external command.  The following escapes are allowed in the
command string:
.nf

    %f => PDF file name (or an empty string if no
          file is open)
    %b => PDF file base name, i.e., file name minus
          the extension (or an empty string if no
          file is open)
    %u => link URL (or an empty string if not over
          a URL link)
    %p => current page number (or an empty string if
          no file is open)
    %x => selection upper-left x coordinate
          (or 0 if there is no selection)
    %y => selection upper-left y coordinate
          (or 0 if there is no selection)
    %X => selection lower-right x coordinate
          (or 0 if there is no selection)
    %Y => selection lower-right y coordinate
          (or 0 if there is no selection)
    %i => page containing the mouse pointer
    %j => x coordinate of the mouse pointer
    %k => y coordinate of the mouse pointer
    %% => %

.fi
The external command string will often contain spaces, so the whole
command must be quoted in the xpdfrc file:
.nf

    bind x "run(ls -l)"

.fi
.TP
.B openOutline
Open the outline pane.
.TP
.B closeOutline
Close the outline pane.
.TP
.B toggleOutline
Toggle the outline pane between open and closed.
.TP
.BI scrollOutlineDown( n )
Scroll the outline down by
.I n
increments.
.TP
.BI scrollOutlineUp( n )
Scroll the outline up by
.I n
increments.
.TP
.B focusToDocWin
Set the keyboard focus to the main document window.
.TP
.B focusToPageNum
Set the keyboard focus to the page number text box.
.TP
.B focusToZoom
Set the keyboard focus to the zoom number text box.
.TP
.B find
Open the 'find' dialog.
.TP
.BI search( key )
Search for the string
.I key
as if it had been entered in the 'find' dialog.
.TP
.B findNext
Finds the next occurrence of the search string (no dialog).
.TP
.B findPrev
Finds the previous occurrence of the search string (no dialog).
.TP
.B print
Open the 'print' dialog.
.TP
.B about
Open the 'about' dialog.
.TP
.B quit
Quit from xpdf.
.PP
The following commands depend on the current mouse position:
.TP
.B startSelection
Start a selection, which will be extended as the mouse moves.
.TP
.B endSelection
End a selection.
.TP
.B startPan
Start a pan, which will scroll the document as the mouse moves
.TP
.B endPan
End a pan.
.TP
.B postPopupMenu
Display the popup menu.
.TP
.B followLink
Follow a hyperlink (does nothing if the mouse is not over a link).
.TP
.B followLinkInNewWin
Follow a hyperlink, opening PDF files in a new window (does nothing if
the mouse is not over a link).  For links to non-PDF files, this
command is identical to followLink.
.TP
.B followLinkNoSel
Same as followLink, but does nothing if there is a non-empty selection.
(This is useful as a mouse button binding.)
.TP
.B followLinkInNewWinNoSel
Same as followLinkInNewWin, but does nothing if there is a non-empty
selection.  (This is useful as a mouse button binding.)
.SS Default Bindings
The default mouse bindings are as follows:
.nf

    bind mousePress1    any         startSelection
    bind mouseRelease1  any         endSelection followLinkNoSel
    bind mousePress2    any         startPan
    bind mouseRelease2  any         endPan
    bind mousePress3    any         postPopupMenu
    bind mousePress4    any         scrollUpPrevPage(16)
    bind mousePress5    any         scrollDownNextPage(16)
    bind mousePress6    any         scrollLeft(16)
    bind mousePress7    any         scrollRight(16)

.fi
The default key bindings are as follows:
.nf

    bind ctrl-home      any         gotoPage(1)
    bind home           any         scrollToTopLeft
    bind ctrl-end       any         gotoLastPage
    bind end            any         scrollToBottomRight
    bind pgup           any         pageUp
    bind backspace      any         pageUp
    bind delete         any         pageUp
    bind pgdn           any         pageDown
    bind space          any         pageDown
    bind left           any         scrollLeft(16)
    bind right          any         scrollRight(16)
    bind up             any         scrollUp(16)
    bind down           any         scrollDown(16)
    bind [              any         rotateCCW
    bind ]              any         rotateCW
    bind o              any         open
    bind O              any         open
    bind r              any         reload
    bind R              any         reload
    bind s              any         saveAs
    bind S              any         saveAs
    bind f              any         find
    bind F              any         find
    bind ctrl-f         any         find
    bind /              any         find
    bind ctrl-g         any         findNext
    bind ctrl-v         any         findPrev
    bind ctrl-p         any         print
    bind n              scrLockOff  nextPage
    bind N              scrLockOff  nextPage
    bind n              scrLockOn   nextPageNoScroll
    bind N              scrLockOn   nextPageNoScroll
    bind p              scrLockOff  prevPage
    bind P              scrLockOff  prevPage
    bind p              scrLockOn   prevPageNoScroll
    bind P              scrLockOn   prevPageNoScroll
    bind v              any         goForward
    bind V              any         goForward
    bind b              any         goBackward
    bind B              any         goBackward
    bind g              any         focusToPageNum
    bind G              any         focusToPageNum
    bind 0              any         zoomPercent(125)
    bind +              any         zoomIn
    bind -              any         zoomOut
    bind z              any         zoomFitPage
    bind Z              any         focusToZoom
    bind w              any         zoomFitWidth
    bind W              any         zoomFitWidth
    bind h              any         zoomFitHeight
    bind H              any         zoomFitHeight
    bind alt-f          any         toggleFullScreenMode
    bind ctrl-l         any         redraw
    bind ctrl-w         any         closeWindowOrQuit
    bind ?\&              any         about
    bind q              any         quit
    bind Q              any         quit
    bind escape         any         quit

.fi
Previous versions of xpdf included a "viKeys" X resource.  It is no
longer available, but the following bindings are equivalent:
.nf

    bind h any scrollLeft(16)
    bind l any scrollRight(16)
    bind k any scrollUp(16)
    bind j any scrollDown(16)

.fi
.SH "REMOTE SERVER MODE"
Xpdf can be started in remote server mode by specifying a server name
(in addition to the file name and page number).  For example:
.PP
.RS
xpdf \-remote myServer file.pdf
.RE
.PP
If there is currently no xpdf running in server mode with the name 'myServer',
a new xpdf window will be opened.  If another command:
.PP
.RS
xpdf \-remote myServer another.pdf 9
.RE
.PP
is issued, a new copy of xpdf will not be started.  Instead, the first
xpdf (the server) will load
.I another.pdf
and display page nine.  If the file name is the same:
.PP
.RS
xpdf \-remote myServer another.pdf 4
.RE
.PP
the xpdf server will simply display the specified page.
.PP
The \-raise option tells the server to raise its window; it can be
specified with or without a file name and page number.
.PP
The \-quit option tells the server to close its window and exit.
.SH EXIT CODES
Xpdf uses the following exit codes:
.TP
0
No error.
.TP
1
Error opening a PDF file.
.TP
99
Other error.
.SH AUTHORS
The original Xpdf software and documentation are copyright 1996\(en2014 Glyph &
Cog, LLC.  Modifications for Poppler support and further development were done
by the Debian and Gentoo Xpdf maintainers, and by xpopple contributors.
.SH "SEE ALSO"
.ad l
.BR pdftops (1),
.BR pdftotext (1),
.BR pdftohtml (1),
.BR pdfinfo (1),
.BR pdffonts (1),
.BR pdfdetach (1),
.BR pdftoppm (1),
.BR pdftopng (1),
.BR pdfimages (1),
.BR xpdfrc (5)


xpdf/xpdf.cc
//========================================================================
//
// xpdf.cc
//
// Copyright 1996-2003 Glyph & Cog, LLC
// Copyright 2014-2020 Adam Sampson <ats@offog.org>
//
//========================================================================

#include <poppler-config.h>
#include <memory>
#include <string>
#include <goo/GooString.h>
#include "parseargs.h"
#include <goo/gfile.h>
#include <goo/gmem.h>
#include "GlobalParams.h"
#include "XPDFParams.h"
#include "Object.h"
#include "XPDFApp.h"
#include "config.h"

//------------------------------------------------------------------------
// command line options
//------------------------------------------------------------------------

static bool contView = false;
static char pageCmdArg[256] = "";
static char psFileArg[256];
static char paperSize[15] = "";
static int paperWidth = 0;
static int paperHeight = 0;
static bool level1 = false;
static char textEncName[128] = "";
static char ownerPasswordArg[33] = "\001";
static char userPasswordArg[33] = "\001";
static bool fullScreen = false;
static char remoteName[100] = "xpdf_";
static char remoteCmd[512] = "";
static bool doRemoteReload = false;
static bool doRemoteRaise = false;
static bool doRemoteQuit = false;
static bool printCommands = false;
static bool quiet = false;
static char cfgFileName[256] = "";
static bool printVersion = false;
static bool printHelp = false;

static ArgDesc argDesc[] = {
  {"-g",          argStringDummy, NULL,           0,
   "initial window geometry"},
  {"-geometry",   argStringDummy, NULL,           0,
   "initial window geometry"},
  {"-title",      argStringDummy, NULL,           0,
   "window title"},
  {"-cmap",       argFlagDummy,   NULL,           0,
   "install a private colormap"},
  {"-rgb",        argIntDummy,    NULL,           0,
   "biggest RGB cube to allocate (default is 5)"},
  {"-rv",         argFlagDummy,   NULL,           0,
   "reverse video"},
  {"-papercolor", argStringDummy, NULL,           0,
   "color of paper background"},
  {"-mattecolor", argStringDummy, NULL,           0,
   "color of background outside actual page"},
  {"-z",          argStringDummy, NULL,           0,
   "initial zoom level (percent, 'page', 'width')"},
  {"-cont",       argFlag,        &contView,      0,
   "start in continuous view mode" },
  {"-pagecmd",    argString,      pageCmdArg,     sizeof(pageCmdArg),
   "command to execute on page changes" },
  {"-ps",         argString,      psFileArg,      sizeof(psFileArg),
   "default PostScript file name or command"},
  {"-paper",      argString,      paperSize,      sizeof(paperSize),
   "paper size (letter, legal, A4, A3, match)"},
  {"-paperw",     argInt,         &paperWidth,    0,
   "paper width, in points"},
  {"-paperh",     argInt,         &paperHeight,   0,
   "paper height, in points"},
  {"-level1",     argFlag,        &level1,        0,
   "generate Level 1 PostScript"},
  {"-enc",        argString,      textEncName,    sizeof(textEncName),
   "output text encoding name"},
  {"-opw",        argString,      ownerPasswordArg, sizeof(ownerPasswordArg),
   "owner password (for encrypted files)"},
  {"-upw",        argString,      userPasswordArg, sizeof(userPasswordArg),
   "user password (for encrypted files)"},
  {"-fullscreen", argFlag,        &fullScreen,    0,
   "run in full-screen (presentation) mode"},
  {"-remote",     argString,      remoteName + 5, sizeof(remoteName) - 5,
   "start/contact xpdf remote server with specified name"},
  {"-exec",       argString,      remoteCmd,      sizeof(remoteCmd),
   "execute command on xpdf remote server (with -remote only)"},
  {"-reload",     argFlag,        &doRemoteReload, 0,
   "reload xpdf remote server window (with -remote only)"},
  {"-raise",      argFlag,        &doRemoteRaise, 0,
   "raise xpdf remote server window (with -remote only)"},
  {"-quit",       argFlag,        &doRemoteQuit,  0,
   "kill xpdf remote server (with -remote only)"},
  {"-cmd",        argFlag,        &printCommands, 0,
   "print commands as they're executed"},
  {"-q",          argFlag,        &quiet,         0,
   "don't print any messages or errors"},
  {"-cfg",        argString,      cfgFileName,    sizeof(cfgFileName),
   "configuration file to use in place of .xpdfrc"},
  {"-v",          argFlag,        &printVersion,  0,
   "print copyright and version info"},
  {"-h",          argFlag,        &printHelp,     0,
   "print usage information"},
  {"-help",       argFlag,        &printHelp,     0,
   "print usage information"},
  {"--help",      argFlag,        &printHelp,     0,
   "print usage information"},
  {"-?",          argFlag,        &printHelp,     0,
   "print usage information"},
  {"-aa",         argObsolete1,   NULL,           0,
   NULL},
  {"-aaVector",   argObsolete1,   NULL,           0,
   NULL},
  {"-eol",        argObsolete1,   NULL,           0,
   NULL},
  {"-freetype",   argObsolete1,   NULL,           0,
   NULL},
  {"-t1lib",      argObsolete1,   NULL,           0,
   NULL},
  {NULL}
};

//------------------------------------------------------------------------

int main(int argc, char *argv[]) {
  std::unique_ptr<XPDFApp> app;
  std::unique_ptr<std::string> fileName;
  int pg = 1;
  std::unique_ptr<std::string> destName;
  std::unique_ptr<std::string> userPassword, ownerPassword;
  bool ok;

  // parse args
  ok = parseArgs(argDesc, &argc, argv);
  if (!ok || printVersion || printHelp) {
    fprintf(stderr, "xpdf version %s\n", xpdfVersion);
    fprintf(stderr, "%s\n", xpdfCopyright);
    if (!printVersion) {
      printUsage("xpdf", "[<PDF-file> [<page> | +<dest>]]", argDesc);
    }
    return 99;
  }

  // read config file
#ifdef GLOBALPARAMS_UNIQUE_PTR
  globalParams = std::make_unique<GlobalParams>();
#else
  auto globalParamsPtr = std::make_unique<GlobalParams>();
  globalParams = globalParamsPtr.get();
#endif
  xpdfParams = std::make_unique<XPDFParams>(cfgFileName);
  globalParams->setupBaseFonts(NULL);
  if (contView) {
    xpdfParams->setContinuousView(contView);
  }
  if (pageCmdArg[0]) {
    xpdfParams->setPageCommand(pageCmdArg);
  }
  if (psFileArg[0]) {
    xpdfParams->setPSFile(psFileArg);
  }
  if (paperSize[0]) {
    if (!xpdfParams->setPSPaperSize(paperSize)) {
      fprintf(stderr, "Invalid paper size\n");
    }
  } else {
    if (paperWidth) {
      xpdfParams->setPSPaperWidth(paperWidth);
    }
    if (paperHeight) {
      xpdfParams->setPSPaperHeight(paperHeight);
    }
  }
  if (level1) {
    xpdfParams->setPSLevel(psLevel1);
  }
  if (textEncName[0]) {
    globalParams->setTextEncoding(textEncName);
  }
  if (printCommands) {
    globalParams->setPrintCommands(printCommands);
  }
  if (quiet) {
    globalParams->setErrQuiet(quiet);
  }

  // create the XPDFApp object
  app = std::make_unique<XPDFApp>(&argc, argv);

  // the initialZoom parameter can be set in either the config file or
  // as an X resource (or command line arg)
  if (app->getInitialZoom() != "") {
    xpdfParams->setInitialZoom(app->getInitialZoom());
  }

  // check command line
  ok = ok && argc >= 1 && argc <= 3;
  if (remoteCmd[0]) {
    ok = ok && remoteName[5] && !doRemoteReload && !doRemoteRaise &&
         !doRemoteQuit && argc == 1;
  }
  if (doRemoteReload) {
    ok = ok && remoteName[5] && !doRemoteQuit && argc == 1;
  }
  if (doRemoteRaise) {
    ok = ok && remoteName[5] && !doRemoteQuit;
  }
  if (doRemoteQuit) {
    ok = ok && remoteName[5] && argc == 1;
  }
  if (!ok || printVersion || printHelp) {
    fprintf(stderr, "xpdf version %s\n", xpdfVersion);
    fprintf(stderr, "%s\n", xpdfCopyright);
    if (!printVersion) {
      printUsage("xpdf", "[<PDF-file> [<page> | +<dest>]]", argDesc);
    }
    return 99;
  }
  if (argc >= 2) {
    fileName = std::make_unique<std::string>(argv[1]);
  }
  if (argc == 3) {
    if (argv[2][0] == '+') {
      destName = std::make_unique<std::string>(&argv[2][1]);
    } else {
      pg = atoi(argv[2]);
      if (pg < 0) {
	fprintf(stderr, "Invalid page number (%d)\n", pg);
	return 99;
      }
    }
  }

  // handle remote server stuff
  if (remoteName[5]) {
    app->setRemoteName(remoteName);
    if (app->remoteServerRunning()) {
      if (fileName) {
	if (destName) {
	  app->remoteOpenAtDest(*fileName, *destName, doRemoteRaise);
	} else {
	  app->remoteOpen(*fileName, pg, doRemoteRaise);
	}
      } else if (remoteCmd[0]) {
	app->remoteExec(remoteCmd);
      } else if (doRemoteReload) {
	app->remoteReload(doRemoteRaise);
      } else if (doRemoteRaise) {
	app->remoteRaise();
      } else if (doRemoteQuit) {
	app->remoteQuit();
      }
      return 0;
    }
    if (doRemoteQuit) {
      return 0;
    }
  }

  // set options
  app->setFullScreen(fullScreen);

  // check for password string(s)
  if (ownerPasswordArg[0] != '\001') {
    ownerPassword = std::make_unique<std::string>(ownerPasswordArg);
  }
  if (userPasswordArg[0] != '\001') {
    userPassword = std::make_unique<std::string>(userPasswordArg);
  }

  // open the file and run the main loop
  if (!app->open(fileName.get(), pg, destName.get(),
		 ownerPassword.get(), userPassword.get())) {
    return 1;
  }
  app->run();

  return 0;
}
